README.md
[![Work in Repl.it](https://classroom.github.com/assets/work-in-replit-14baed9a392b3a25080506f3b7b6d57f295ec2978f6f33ec97e36a161684cbe9.svg)](https://classroom.github.com/online_ide?assignment_repo_id=433779&assignment_repo_type=GroupAssignmentRepo)

# Trabalho 03: O Jogo Keno
Nesse trabalho você vai desenvolver um pequeno jogo no terminal que simula um clássico [jogo de azar](https://pt.wikipedia.org/wiki/Jogo_de_azar) chamado [Keno](https://en.wikipedia.org/wiki/Keno).
Mais detalhes sobre o projeto você pode encontrar [aqui](docs/projeto_keno.pdf).

Não esqueça de preencher o arquivo [`autor.md`](author.md) antes de submeter seu projeto.

# Compilando e Executando

## Manualmente com g++

Para compilar e executar o programa manualmente basta digitar os seguintes comandos em seu terminal:

```
$ mkdir bin (caso não tenha uma pasta para os executáveis, você deve criá-la com esse comando)
$ g++ -Wall -std=c++11 -g source/src/main.cpp source/src/keno_bet.cpp source/src/game_manager.cpp -I source/include -o bin/keno (compilar)
$ ./bin/keno <diretório do arquivo com sua aposta> (executar)
```

## Cmake

Para compilar e executar o programa com o Cmake você precisa ter instalado pelo menos a versão 3.2. Em seguida, abra seu terminal e execute os seguintes comandos:

```
$ cmake -S source -B build (pede ao cmake para criar o diretório de compilação)
$ cmake --build build (compilar)
$ ./build/keno <diretório do arquivo com sua aposta> (executar)
```

Os arquivos com apostas devem ser salvos na pasta `data` (se isso for feito, para executar basta `./build/keno ./data/<arquivo com sua aposta>`. Já existem alguns exemplos de arquivos de aposta nesta pasta. É possível utilizá-los, mas você pode criar o seu próprio também.

source/CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/) 
and this project adheres to [Semantic Versioning](http://semver.org/).

## [Unreleased]

# [0.2.1] - 2021-09-25
### Added
- A new welcome message
### Changed
- Architecture (functions and organization)
- Game states
- The old welcome message has become a summary message
- Compiling and running on the `README.md` file

# [0.2.0] - 2021-07-05
### Added
- We can use CMake now
### Changed
- `payout_table` type (const float, now)


# [0.1.2] - 2021-07-05
### Added
- Sorting of sport
- Removing of spots repetitions
### Removed
- Binary file from `bin` directory to do not make the repository large

# [0.1.1] - 2021-07-03
### Added
- Payout table
- STL algorithm header
### Changed
- Fix stringstream
- Try to remove repetitions
### Removed
- `core` directory.

# [0.1.0] - 2021-07-03
### Added
- Make the bet input code
- New files: `keno_common.cpp`, `game_manager.cpp`, and `game_manager.h`.
- `keno_common.cpp` is a file of implementations of the `main.cpp`'s functions.
- New `game_manager.*` files for our new class GameManager (responsible to manager the game).
- New `bin` directory for binary files.
- New `include` directory for header files.
- New `src` directory for implementation files.
### Changed
- Move files from `core` directory to `include` and `src` directories.


source/src/keno_bet.cpp
/*!
 * @brief Class that defines a Keno bet.
 *
 * @author João Pedro Fonseca Dantas
 * @author João Guilherme Alves Lopes da Costa
 * @file keno_bet.cpp
 */

#include "keno_bet.h"

namespace keno {
  
  bool KenoBet::AddSpot(keno::number_type spot) {
    if (binary_search(KenoBet::m_spots.begin(), KenoBet::m_spots.end(), spot)) {
      return false;
    } else {
      KenoBet::m_spots.push_back(spot);
      return true;
    }
  }

  bool KenoBet::set_wage(keno::cash_type wage) {
    if (wage <= 0) return false;
    else KenoBet::m_wage = wage;
    return true;
      // spots.erase(last, spots.end());
  }

  void KenoBet::reset(void) {
    KenoBet::m_spots.erase(KenoBet::m_spots.begin(), KenoBet::m_spots.end());
    KenoBet::m_wage = 0;
  }

  keno::cash_type KenoBet::get_wage(void) const {
    return m_wage;
  }

  size_t KenoBet::Size(void) const {
    return m_spots.size();
  }

  keno::set_of_numbers_type
  KenoBet::FindHits(const keno::set_of_numbers_type &draw) const {
    // This container will store all common elements, if any exists.
    keno::set_of_numbers_type intersection;

    for (unsigned short i = 0; i < draw.size(); i++) {
      if (binary_search(KenoBet::m_spots.begin(),
                        KenoBet::m_spots.end(),
                        draw[i])) {
        intersection.push_back(draw[i]);
      }
    }

    return intersection;
  }

  keno::set_of_numbers_type
  KenoBet::get_spots(void) const {
    return m_spots;
  }

}



source/include/keno_bet.h
#ifndef _KENO_BET_H_
#define _KENO_BET_H_

/*!
 * @brief Class that defines a Keno bet.
 *
 * @author Selan R. dos Santos
 * @author João Pedro Fonseca Dantas
 * @author João Guilherme Lopes Alves da Costa
 * @file keno_bet.h 
 */

#include "keno_common.h"

/*!
 * @namespace Keno
 */
namespace keno {  
  class KenoBet {
    public:

      /*!
       * @brief Creates a Keno bet.
       * A Keno bet stores the player's picked numbers or spots, up to a limit passed as argument.
       * Any attempt of inserting spots beyond the set limit is ignored.
       * @param max_n_sposts_ Maximim number of spots this bet may hold.
       */
      KenoBet( ) : m_wage{0.f}
      { /* empty */ }

      /*! 
       * @brief Adds a number to the player's bet.
       * This method adds the number passed as argument to the player's spots
       * only if the number is not already there.
       * @param spot The number we wish to include in the bet.
       * @return Returns true if number chosen is successfuly inserted; false otherwise.
       */
      bool AddSpot(keno::number_type spot);

      /*!
       * @brief Sets the player's wage on this bet.
       * This method set the amount of money the player is betting.
       * @param wage The wage.
       * @return True if we have a wage above zero; false otherwise.
       */
      bool set_wage(keno::cash_type wage);

      /*!
       * @brief Resets a bet to an empty state.
       * Resets a bet object to an empty state, which means the spot list clered.
       */
      void reset(void);

      /*! 
       * @brief Recovers the player's wage on this particular bet.
       * Retrieves the player's wage on this bet.
       * @return The wage value.
       */
      keno::cash_type get_wage(void) const;


      /*!
       * @brief Return the amount of spots currently in the bet.
       * Returns to the application the current number of spots in the player's bet.
       * @return Number of spots present in the bet.
       */
      size_t Size(void) const;

      /*!
       * @brief Determine how many spots from the player's match the hits passed as argument.
       * @param draw List of hits randomly chosen by the computer.
       * @return An vector with the list of hits according to the currently stored bet.
       */
      keno::set_of_numbers_type
      FindHits(const keno::set_of_numbers_type &draw) const;

      /*! 
       * @brief Return the player's spots.
       * Return an std::vector<number_type> with the spots the player has picked so far.
       * @return The std::vector<number_type> with the plyer's spots picked so far.
       */
      keno::set_of_numbers_type get_spots(void) const;

    private:
      keno::set_of_numbers_type m_spots; //<! The player's bet, i.e.\ a set of numbers (aka spots).
      keno::cash_type m_wage;            //<! The player's wage
  };
}

#endif


source/include/keno_common.h
#ifndef _KENO_COMMON_H_
#define _KENO_COMMON_H_

/*!
 * @brief This is a set of common definition for classes GameManager and KenoBet.
 * This is to avoid circular referencing.
 *
 * @author Selan R. dos Santos
 * @author João Pedro Fonseca Dantas 
 * @author João Guilherme Alves Lopes da Costa
 * @file keno_common.h
 */

#include <fstream> ///< To handling files.
#include <sstream> ///< To handling the stringstream.
#include <iostream> ///< To use cout and cerr.
#include <vector> ///< To use vector and its methods.
#include <string> ///< To use string ans its methods.
#include <random> ///< To shuffle and random.
#include <algorithm> ///< To sort and get uniques.

namespace keno {
  using number_type = unsigned short int; ///< Data type for a keno hit.
  using cash_type = float; ///< Defines the wage type in this application.
  using set_of_numbers_type = std::vector<number_type>; ///< List of numbers for a keno bet.

  const float PAYOUT_TABLE[15][16] = {
    {0,   3,  -1,  -1,  -1, -1, -1,  -1,  -1,   -1,   -1,   -1,   -1,   -1,   -1,    -1},
    {0,   1,   9,  -1,  -1, -1, -1,  -1,  -1,   -1,   -1,   -1,   -1,   -1,   -1,    -1},
    {0,   1,   2,  16,  -1, -1, -1,  -1,  -1,   -1,   -1,   -1,   -1,   -1,   -1,    -1},
    {0, 0.5,   2,   6,  12, -1, -1,  -1,  -1,   -1,   -1,   -1,   -1,   -1,   -1,    -1},
    {0, 0.5,   1,   3,  15, 50, -1,  -1,  -1,   -1,   -1,   -1,   -1,   -1,   -1,    -1},
    {0, 0.5,   1,   2,   3, 30, 75,  -1,  -1,   -1,   -1,   -1,   -1,   -1,   -1,    -1},
    {0, 0.5, 0.5,   1,   6, 12, 36, 100,  -1,   -1,   -1,   -1,   -1,   -1,   -1,    -1},
    {0, 0.5, 0.5,   1,   3,  6, 19,  90, 720,   -1,   -1,   -1,   -1,   -1,   -1,    -1},
    {0, 0.5, 0.5,   1,   2,  4,  8,  20,  80, 1200,   -1,   -1,   -1,   -1,   -1,    -1},
    {0,   0, 0.5,   1,   2,  3,  5,  10,  30,  600, 1800,   -1,   -1,   -1,   -1,    -1},
    {0,   0, 0.5,   1,   1,  2,  6,  15,  25,  180, 1000, 3000,   -1,   -1,   -1,    -1},
    {0,   0,   0, 0.5,   1,  2,  4,  24,  72,  250,  500, 2000, 4000,   -1,   -1,    -1},
    {0,   0,   0, 0.5, 0.5,  3,  4,   5,  20,   80,  240,  500, 3000, 6000,   -1,    -1},
    {0,   0,   0, 0.5, 0.5,  2,  3,   5,  12,   50,  150,  500, 1000, 2000, 7500,    -1},
    {0,   0,   0, 0.5, 0.5,  1,  2,   5,  15,   50,  150,  300,  600, 1200, 2500, 10000}
  }; ///< Table of payout factors values.
  const cash_type MAX_WAGE{100000}; ///< Maximum number of a wage.

  /**
   * @brief Enumeration to determine the states of the program.
   */
  enum State {
    START = 0, //!< The initial state.
    END, //!< Finishing the program.
    WELCOME, //!< Welcome message.
    READING, //!< Reading the input file.
    PLAYING, //!< Animating the bar charts.
  };

  /**
   * @brief Enumeration to identify a error or warning.
   */
  enum Error {
    kInputManyArguments = 0, //!< The input has too many argumens.
    kMaxSpotsExceeded, //!< Player entered with too many spots.
    kInputFewArguments, //!< The input has too few arguments.
    kBetFileNotFound, //!< Could not find the bet file.
    kInvalidBet, //!< Player entered a invalid bet.
    kInvalidRounds, //!< Player entered a invalid number of rounds.
    kSpotsNotInformed, //!< Player did not inform its spots.
  };
};

#endif


source/include/game_manager.h
#ifndef _GAME_MANAGER_H_
#define _GAME_MANAGER_H_

/*!
 * @brief Class that defines the management of the Keno game.
 *
 * @author João Pedro Fonseca Dantas
 * @author João Guilherme Alves Lopes da Costa
 * @file game_manager.cpp
 */

#include "keno_common.h"
#include "keno_bet.h"

namespace keno { 
  /**
   * @class
   */
  class GameManager {
    public:
      
      /**
       * @brief Write to user some error or warning found in program process.
       * @param type Error or warning found.
       */
      void WriteError(keno::Error type);
      
      /**
       * @brief Display a welcome message.
       */
      void DisplayWelcome(void);

      /**
       * @brief Display the information read from the data file and a guide.
       */
      void DisplaySummary(void);
      
      /**
       * @brief Display the result of the game.
       */
      void DisplayEnd(void);

      /**
       * @brief Read bet data of a given bet file. 
       * @param path Given bet file path.
       */
      void ReadBetFile(std::string &path);

      /**
       * @brief Start a new game from a given input.
       * @param argc Number of arguments (inputs) of the program execution.
       * @param argv Pointer to the given arguments (inputs).
       */
      void Initialize(int argc, char *argv[]);

      /**
       * @brief Check if the program is still running.
       * @return true if the program has finished.
       * @return false otherwise.
       */
      bool GameOver(void);

      /**
       * @brief Generate a draw and find hits.
       */
      void ProcessEvents(void);

      /**
       * @brief Update the current state of the program.
       */
      void Update(void);

      /**
       * @brief Display the current state of the program after the processing.
       */
      void Render(void);

    private:
      keno::cash_type cash_per_round; ///< The cash to wage for each round
      keno::cash_type initial_credit; ///< Player credit at the beggining of the game
      keno::KenoBet player_bet; ///< The bet given by the player
      keno::number_type current_round{0}; ///< Number of the current round
      keno::number_type number_of_rounds; ///< Total number of rounds
      keno::set_of_numbers_type draw; ///< List of drawn numbers
      keno::set_of_numbers_type hits; ///< Intersection between drawn numbers and player spots
      keno::set_of_numbers_type spots; ///< Chosen number by the player
      keno::State game_state; ///< Representation of error or warning
      std::string bet_file_path; ///< String of the path to the player bet file 
  };
}

#endif


source/src/game_manager.cpp
/*!
 * @brief Class that defines the management of the Keno game.
 *
 * @author João Pedro Fonseca Dantas
 * @author João Guilherme Alves Lopes da Costa
 * @file game_manager.cpp
 */

#include "game_manager.h"
#include "keno_common.h"

/*!
 * @namespace Keno
 */
namespace keno {
  //============[ GameManager METHODS ]===============//

  void GameManager::WriteError(keno::Error type) {
    switch (type) {
      case kInputManyArguments:
        std::cout
          << "\n>>> WARNING! Many arguments to run the game.\n"
          << "    You only need to specify the executable file and one data input file.\n"
          << "    The arguments after the first input file will be ignored.\n"
          << "    Correct example: ./bin/keno ./data/bet_01.dat\n";
          break;
      case kMaxSpotsExceeded:
        std::cerr
          << "\n>>> WARNING! You can only enter a maximum of 15 numbers for spots.\n"
          << "    We are considering only the first 15 unique valid numbers.\n\n";
          break;
      case kInputFewArguments:
        std::cerr
          << "\n>>> ERROR! You did not specify the data input file to bet.\n"
          << "    You need to run the game and enter the bet data via a file.\n"
          << "    Example: ./bin/keno ./data/bet_01.dat\n\n";
        break;
      case kBetFileNotFound:
        std::cerr
          << "\n>>> ERROR! We didn't can read the file [" << GameManager::bet_file_path << "]\n"
          << "    Probably [" << GameManager::bet_file_path << "] is not a data file or don't exist.\n\n";
          break;
      case kInvalidBet:
        std::cerr
          << "\n>>> ERROR! You have not entered the amount you want to bet or your amount is not allowed.\n"
          << "    Next time enter a value greater than 1 and less than 100000 for your bet.\n\n";
          break;
      case kInvalidRounds:
        std::cerr
          << "\n>>> ERROR! You did not enter the number of rounds or entered a non-positive value.\n"
          << "    Next time enter a value greater than 0 for the number of rounds.\n\n";
          break;
      case kSpotsNotInformed:
        std::cerr
          << "\n>>> ERROR! You did not enter the spots you would like to bet.\n"
          << "    Next time enter at least one spot to play the game.\n\n";
          break;
      default:
        std::cerr << "\n>>> UNKNOWN ERROR!\n";
    }
    if (type > 1) exit(1);
  }

  void GameManager::DisplayWelcome(void) {
    std::cout << "===================================================\n";
    std::cout << "          Welcome to the Keno Game, v2.1\n";
    std::cout << "   Copyright © 2021, João Guilherme & João Pedro\n";
    std::cout << "===================================================\n";
  }

  void GameManager::DisplaySummary(void) {
    keno::set_of_numbers_type numbers = player_bet.get_spots();
    std::cout
      << "    You are going to wage a total of $" << GameManager::initial_credit << " dollars.\n"
      << "    Going for a total of " << GameManager::number_of_rounds << " rounds, waging $" << cash_per_round << " per round.\n\n"
      << "    Your bet has " << player_bet.Size() << " numbers. They are: [ ";
    for (number_type i{0}; i < player_bet.Size(); i++) {
      std::cout << numbers[i] << " ";
    }

    std::cout
      << "]\n"
      << "         -------+---------\n"
      << "         Hits   | Payout\n"
      << "         -------+---------\n";
    for (number_type i{0}; i <= player_bet.Size(); i++) {
      std::cout << "         " << i;
      if (i < 10) {
        std::cout << "      | ";
      }
      else {
        std::cout << "     | ";
      }
      std::cout << keno::PAYOUT_TABLE[player_bet.Size() - 1][i];
      if (i != player_bet.Size()) {
        std::cout << "\n";
      }
    }
  }

  void GameManager::DisplayEnd(void) {
    std::cout
      << "\n>>> End of rounds!\n"
      << "--------------------------------------------------\n\n"
      << "===== SUMMARY =====\n"
      << ">>> You spent in this game a total of $" << initial_credit << " dollars.\n";
    
    keno::cash_type amount{player_bet.get_wage() - initial_credit};
    if (amount > 0) {
      std::cout << ">>> Hooray, you won $" << std::abs(amount) << " dollars. See you next time! ;-)\n";
    }
    else if (amount < 0) {
      std::cout << ">>> Oh no, you lost $" << std::abs(amount) << " dollars. More luck next time! :-(\n";
    }
    else {
      std::cout << ">>> Oops, you kept your initial bet money, you didn't win or lose. More luck next time! :-/\n";
    }
    std::cout << ">>> You are leaving the Keno table with $" << player_bet.get_wage() << " dollars.\n";
  }

  void GameManager::ReadBetFile(std::string &path) {
    std::string line;
    std::stringstream ss;
    keno::number_type number;
    std::ifstream bet_file;
    GameManager::spots.resize(0);

    bet_file.open(path);
    std::cout
      << "\n>>> Preparing to read bet file [" << bet_file_path << "], please wait...\n"
      << "--------------------------------------------------\n";
    if (not bet_file.is_open())
      GameManager::WriteError(kBetFileNotFound);
    std::getline(bet_file, line);
    if (bet_file.fail()) {
      GameManager::WriteError(kBetFileNotFound);
    }
    bet_file.seekg(0); // Put pointer at the beginning of the file
    
    bool get_input{false};
    keno::cash_type player_credit{0};
    while (not bet_file.eof() && not get_input) {
      std::getline(bet_file, line);
      ss.clear();
      ss << line;
      if (ss >> player_credit) {
        if (line[0] == '-' || player_credit > MAX_WAGE) {
          GameManager::WriteError(kInvalidBet);
        }
        get_input = true;
      }
    }
    GameManager::initial_credit = player_credit;

    get_input = false;
    keno::number_type nr{0};
    while (not bet_file.eof() && not get_input) {
      std::getline(bet_file, line);
      ss.clear();
      ss << line;
      if (ss >> nr) {
        if (line[0] == '-' || nr == 0) {
          GameManager::WriteError(kInvalidRounds);
        }
        get_input = true;
      }
    }
    GameManager::number_of_rounds = nr;
    GameManager::cash_per_round = initial_credit / number_of_rounds;

    get_input = false;
    keno::set_of_numbers_type player_numbers;
    while (not bet_file.eof() && not get_input) {
      std::getline(bet_file, line);
      ss.clear();
      ss << line;
      while (ss >> number) {
        if (player_numbers.size() == 15) {
          GameManager::WriteError(kMaxSpotsExceeded);
          break;
        }
        else
          player_numbers.push_back(number);
      }
      GameManager::spots = player_numbers;
    }
    // std::cout << "bet: " << player_credit << " nr: " << nr << " spots: " << spots.size() << std::endl;
    if (player_credit == 0) {
      GameManager::WriteError(kInvalidBet);
    }
    else if (nr == 0) {
      GameManager::WriteError(kInvalidRounds);
    }
    else if (spots.size() == 0) {
      GameManager::WriteError(kSpotsNotInformed);
    }
    std::sort(spots.begin(), spots.end());

    bet_file.close();
    std::cout << ">>> Bet successfully read!\n";
  }
  
  void GameManager::Initialize(int argc, char *argv[]) {
    if (argc > 2)
      GameManager::WriteError(kInputManyArguments);
    else if (argc < 2)
      GameManager::WriteError(kInputFewArguments);
    
    GameManager::bet_file_path = argv[1];
    game_state = keno::State::START;
  }

  bool GameManager::GameOver(void) {
    if (game_state == keno::State::END) {
      player_bet.reset();
      return true;
    }
    else {
      return false;
    }
  }

  void GameManager::ProcessEvents(void) {
    if (game_state == keno::State::WELCOME) {
      // Calls the function that reads the input file.
      GameManager::ReadBetFile(bet_file_path);
      for (keno::number_type i{0}; i < spots.size(); i++) {
          player_bet.AddSpot(spots[i]);
      }
      player_bet.set_wage(initial_credit);
    }
    else if(game_state == keno::State::READING) {
      // Waits for the user to press enter to start the game.
      while (getchar() != '\n') {}
    }
    else if (game_state == keno::State::PLAYING) {
      // Waits for the user to press enter to start the round.
      while (getchar() != '\n') {}
      // Sortear Números;
      keno::set_of_numbers_type range;
      keno::number_type value{1};
      for (number_type i{0}; i < 80; i++) {
        range.push_back(value);
        value++;
      }
      std::random_device rd;  // Will be used to obtain a seed for the random number engine
      std::mt19937 gen(rd()); // Standard mersenne_twister_engine seeded with rd()
      std::shuffle(range.begin(), range.end(), gen);
      draw.assign(range.begin(), range.begin()+20);
      // Ordenar números;
      std::sort(draw.begin(), draw.end());
      // Ver interseção (FindHits)
      hits = player_bet.FindHits(draw);
    }
  }

  void GameManager::Update(void) {
    if (game_state == keno::State::START) {
      game_state = keno::State::WELCOME;
    }
    else if (game_state == keno::State::WELCOME) {
      game_state = keno::State::READING;
    }
    else if (game_state == keno::State::READING) {
      game_state = keno::State::PLAYING;
    }
    else if (game_state == keno::State::PLAYING) {
      if (current_round != number_of_rounds) {
        keno::cash_type total_payout = PAYOUT_TABLE[player_bet.Size() - 1][hits.size()] * cash_per_round;
        keno::cash_type cash_round = total_payout - cash_per_round;
        keno::cash_type player_wage = player_bet.get_wage();
        player_bet.set_wage(player_wage + cash_round);
        current_round++;
      }
      else {
        game_state = keno::State::END;
      }
    }
  }

  void GameManager::Render(void) {
    if (game_state == keno::State::WELCOME) {
      // Display a welcoming message.
      DisplayWelcome();
    }
    else if (game_state == keno::State::READING) {
      // Display a summary of information captured from the bet file.
      DisplaySummary();
    }
    else if (game_state == keno::State::PLAYING) {
      std::cout 
        << "\n" << "         --------------------------------------------------\n"
        << "         This is round #" << current_round << " of " << number_of_rounds << " and your wage is $" << cash_per_round << ". Good luck!\n"
        << "         The draw numbers are: [ ";
      for (keno::number_type i{0}; i < draw.size(); i++) {
        std::cout << draw[i] << " ";
      }
      std::cout 
        << "]\n\n"
        << "         You hit the following number(s) [ ";
      for (keno::number_type i{0}; i < hits.size(); i++) {
        std::cout << hits[i] << " ";
      }
      std::cout
        << "], a total of " << hits.size() << " hits out of " << player_bet.Size() << ".\n"
        << "         Payout rate is " << PAYOUT_TABLE[player_bet.Size() - 1][hits.size()] << ", thus you came out with: $" 
        << PAYOUT_TABLE[player_bet.Size() - 1][hits.size()] * cash_per_round << "\n"
        << "         Your net balance so far is: $" << player_bet.get_wage() << " dollars.";
    }
    else if (game_state == keno::State::END) {
      DisplayEnd();
    }
  }
  
  //============[ End GameManager class ]===============//

} // namespace Keno


